File: app/src/main/java/com/hellowo/teamfinder/data/TeamsLiveData.java
package com.hellowo.teamfinder.data;

import android.arch.lifecycle.LiveData;
import android.support.annotation.MainThread;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.hellowo.teamfinder.model.Team;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TeamsLiveData extends LiveData<List<Team>> {
    private static TeamsLiveData sInstance;
    private DatabaseReference mDatabase;
    private List<Team> currentList;
    int gameId = -1;

    @MainThread
    public static TeamsLiveData get() {
        if (sInstance == null) {
            sInstance = new TeamsLiveData();
        }
        return sInstance;
    }

    private TeamsLiveData() {
        mDatabase = FirebaseDatabase.getInstance().getReference();
        currentList = new ArrayList<>();
        setValue(currentList);
    }

    public void loadTeams() {
        mDatabase.child(Team.DB_REF)
                .orderByChild(Team.KEY_FILTERING)
                .startAt(System.currentTimeMillis() + "_0")
                .addListenerForSingleValueEvent(
                        new ValueEventListener() {
                            @Override
                            public void onDataChange(DataSnapshot dataSnapshot) {
                                currentList.clear();
                                for (DataSnapshot postSnapshot: dataSnapshot.getChildren()) {
                                    Team team = postSnapshot.getValue(Team.class);
                                    team.setId(postSnapshot.getKey());
                                    currentList.add(0, team);
                                }
                                Collections.sort(currentList, (l,r)->
                                    l.getDtCreated() > r.getDtCreated() ? -1 :
                                            l.getDtCreated() < r.getDtCreated() ? 1 : 0);
                                setValue(currentList);
                            }
                            @Override
                            public void onCancelled(DatabaseError databaseError) {}
                        });
    }

    @Override
    protected void onActive() {
        loadTeams();
    }

}

File: app/src/main/java/com/hellowo/teamfinder/data/TeamsLiveData.kt
package com.hellowo.teamfinder.data

import android.arch.lifecycle.LiveData
import android.support.annotation.MainThread

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.hellowo.teamfinder.model.Team
import com.hellowo.teamfinder.utils.KEY_FILTERING
import com.hellowo.teamfinder.utils.KEY_TEAMS

import java.util.ArrayList
import java.util.Collections

object TeamsLiveData : LiveData<List<Team>>() {
    internal val mDatabase: DatabaseReference = FirebaseDatabase.getInstance().reference
    internal val currentList: MutableList<Team> = ArrayList()

    init {
        value = currentList
    }

    fun loadTeams() {
        mDatabase.child(KEY_TEAMS)
                .orderByChild(KEY_FILTERING)
                .startAt(System.currentTimeMillis().toString() + "_0")
                .addListenerForSingleValueEvent(
                        object : ValueEventListener {
                            override fun onDataChange(dataSnapshot: DataSnapshot) {
                                currentList.clear()

                                for (postSnapshot in dataSnapshot.children) {
                                    val team = postSnapshot.getValue(Team::class.java)
                                    team!!.id = postSnapshot.key
                                    currentList.add(0, team)
                                }

                                Collections.sort(currentList) { l, r ->
                                    if (l.dtCreated > r.dtCreated) -1
                                    else if (l.dtCreated < r.dtCreated) 1
                                    else 0
                                }

                                value = currentList
                            }

                            override fun onCancelled(databaseError: DatabaseError) {}
                        })
    }

    override fun onActive() {
        loadTeams()
    }
}
